File: routing_test.go
// MIT License
//
// Copyright (c) 2026 sparetimecoders
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package spec

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestMatchRoutingKey(t *testing.T) {
	tests := []struct {
		name       string
		pattern    string
		routingKey string
		expected   bool
	}{
		{"exact match", "Order.Created", "Order.Created", true},
		{"no match", "Order.Created", "Order.Deleted", false},
		{"star wildcard matches single", "Order.*", "Order.Created", true},
		{"star wildcard does not match multi", "Order.*", "Order.Created.V2", false},
		{"hash wildcard matches single", "Order.#", "Order.Created", true},
		{"hash wildcard matches multi", "Order.#", "Order.Created.V2", true},
		{"hash wildcard matches everything", "#", "Order.Created", true},
		{"star in middle", "*.Created", "Order.Created", true},
		{"star in middle no match", "*.Created", "Order.Deleted", false},
		{"invalid regex pattern", "[", "anything", false},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			assert.Equal(t, tc.expected, MatchRoutingKey(tc.pattern, tc.routingKey))
		})
	}
}

func TestRoutingKeyToRegex(t *testing.T) {
	tests := []struct {
		input    string
		expected string
	}{
		{"Order.Created", `^Order\.Created$`},
		{"Order.*", `^Order\.[^.]*$`},
		{"Order.#", `^Order\..*$`},
		{"#", `^.*$`},
		{"*.Created", `^[^.]*\.Created$`},
	}

	for _, tc := range tests {
		t.Run(tc.input, func(t *testing.T) {
			assert.Equal(t, tc.expected, routingKeyToRegex(tc.input))
		})
	}
}

func TestRoutingKeyOverlaps(t *testing.T) {
	tests := []struct {
		name     string
		p1       string
		p2       string
		expected bool
	}{
		{"identical patterns", "Order.Created", "Order.Created", true},
		{"no overlap", "Order.Created", "Order.Deleted", false},
		{"wildcard overlaps exact", "Order.#", "Order.Created", true},
		{"exact overlaps wildcard", "Order.Created", "Order.#", true},
		{"star overlaps exact", "Order.*", "Order.Created", true},
		{"non-overlapping wildcards", "Order.*", "User.*", false},
		{"hash overlaps star", "Order.#", "Order.*", true},
		{"invalid regex", "[", "Order.Created", false},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			assert.Equal(t, tc.expected, RoutingKeyOverlaps(tc.p1, tc.p2))
		})
	}
}

File: typescript/__tests__/routing.test.ts
import { describe, expect, it } from "bun:test";
import { matchRoutingKey, routingKeyOverlaps } from "../src/routing.js";

describe("matchRoutingKey", () => {
  const cases: Array<{ name: string; pattern: string; routingKey: string; expected: boolean }> = [
    { name: "exact match", pattern: "Order.Created", routingKey: "Order.Created", expected: true },
    { name: "no match", pattern: "Order.Created", routingKey: "Order.Deleted", expected: false },
    { name: "star wildcard matches single", pattern: "Order.*", routingKey: "Order.Created", expected: true },
    { name: "star wildcard does not match multi", pattern: "Order.*", routingKey: "Order.Created.V2", expected: false },
    { name: "hash wildcard matches single", pattern: "Order.#", routingKey: "Order.Created", expected: true },
    { name: "hash wildcard matches multi", pattern: "Order.#", routingKey: "Order.Created.V2", expected: true },
    { name: "hash wildcard matches everything", pattern: "#", routingKey: "Order.Created", expected: true },
    { name: "star in middle", pattern: "*.Created", routingKey: "Order.Created", expected: true },
    { name: "star in middle no match", pattern: "*.Created", routingKey: "Order.Deleted", expected: false },
    { name: "invalid regex pattern", pattern: "[", routingKey: "anything", expected: false },
  ];

  for (const tc of cases) {
    it(tc.name, () => {
      expect(matchRoutingKey(tc.pattern, tc.routingKey)).toBe(tc.expected);
    });
  }
});

describe("routingKeyOverlaps", () => {
  const cases: Array<{ name: string; p1: string; p2: string; expected: boolean }> = [
    { name: "identical patterns", p1: "Order.Created", p2: "Order.Created", expected: true },
    { name: "no overlap", p1: "Order.Created", p2: "Order.Deleted", expected: false },
    { name: "wildcard overlaps exact", p1: "Order.#", p2: "Order.Created", expected: true },
    { name: "exact overlaps wildcard", p1: "Order.Created", p2: "Order.#", expected: true },
    { name: "star overlaps exact", p1: "Order.*", p2: "Order.Created", expected: true },
    { name: "non-overlapping wildcards", p1: "Order.*", p2: "User.*", expected: false },
    { name: "hash overlaps star", p1: "Order.#", p2: "Order.*", expected: true },
    { name: "invalid regex", p1: "[", p2: "Order.Created", expected: false },
  ];

  for (const tc of cases) {
    it(tc.name, () => {
      expect(routingKeyOverlaps(tc.p1, tc.p2)).toBe(tc.expected);
    });
  }
});
